### PR TITLE
Lots2

### PR DESCRIPTION
**Pull request by AlexRedisson18**
## What was done?
- add category and categories
- add basic validations (index, new, create)
- add conditional validation for interesting categories / price
- change bootstrap select-tag style  to 'bootstrap-select' package style

Screenshot from 2020-06-24 13-03-14
Screenshot from 2020-06-24 13-03-22

### CODE REVIEW

File: db/migrate/20200423070105_create_lots.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@ def change
     create_table :lots do |t|
       t.string :title, null: false
       t.string :description
-      t.string :image, null: false
+      t.string :image
```
Comment (akurashev) [inline db/migrate/20200423070105_create_lots.rb]: Again, changed migration

File: db/schema.rb [hunk 2]
```diff
@@ -10,24 +10,35 @@
...
 
-ActiveRecord::Schema.define(version: 2020_04_23_083056) do
+ActiveRecord::Schema.define(version: 2020_06_18_084818) do
 
   # These are extensions that must be enabled in order to support this database
   enable_extension "plpgsql"
 
   create_table "categories", force: :cascade do |t|
     t.string "name", null: false
+    t.bigint "lot_id"
```
Comment (akurashev) [inline db/schema.rb]: Where is it from?

File: app/models/lot.rb [hunk 3]
```diff
@@ -4,17 +4,27 @@ class Lot < ApplicationRecord
 
   has_many :incoming_offers, class_name: 'Offer', inverse_of: :suggested_lot
   has_many :outgoing_offers, class_name: 'Offer', inverse_of: :requested_lot
-  belongs_to :category
+  has_and_belongs_to_many :interesting_categories, class_name: 'Category', inverse_of: :interested_lot
+  belongs_to :category, optional: true
```
Comment (akurashev) [inline app/models/lot.rb]: Why optional?

File: spec/support/shared_examples/requests/common.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+shared_examples 'response with code' do |code:, request_required: false|
```
Comment (akurashev) [inline spec/support/shared_examples/requests/common.rb]: No need of `request_required` param here. It's always `true`

File: spec/requests/lots_request_spec.rb [hunk 5]
```diff
@@ -1,4 +1,55 @@
...
+
+    context 'when user is NOT signed in' do
+      subject(:make_request) { post :create, params: { lot: lot } }
+
+      let(:category) { create(:category) }
+      let(:lot) { attributes_for(:lot, category_id: category) }
+
+      it_behaves_like 'response with code', code: 302, request_required: true
+    end
+  end
```
Comment (akurashev) [inline spec/requests/lots_request_spec.rb]:
```suggestion
  describe 'POST #create' do
    subject(:make_request) { post :create, params: { lot: lot } }

    let(:category) { create(:category) }
    let(:lot) { attributes_for(:lot, category_id: category) }
    
    context 'when user is signed in' do
      include_context 'with current user'

      it 'creates new lot' do
        expect { make_request }.to change(Lot, :count).from(0).to(1)
      end
    end

    context 'when user is NOT signed in' do
      it_behaves_like 'response with code', code: 302, request_required: true
    end
  end
```

**Review by akurashev — CHANGES_REQUESTED** [review]

File: db/migrate/20200423070105_create_lots.rb (hunk 1 above)
Comment (AlexRedisson18) [reply db/migrate/20200423070105_create_lots.rb]: this is the same change

File: db/schema.rb (hunk 2 above)
Comment (AlexRedisson18) [reply db/schema.rb]: hm...
mb jointable did this?

**Review by mkalinin — APPROVED** [review]
Well done 👍

**Comment by akurashev** [comment]
> Well done 👍

Medium well, please

**Review by akurashev — APPROVED** [review]

**Comment by mkalinin** [comment]
> Medium well, please

lmfao

**Pull request merged**